### DISCUSSION TITLE
Avoid using global variables

### DISCUSSION THREAD

**Discussion by lukejagodzinski** [remix-run/remix#1803] [Q&A]
Hi, I have a question whether you have some pattern to avoid using global variables. Recently, I'm trying to avoid passing global variables around and unfortunately that's the advices way of doing it in Remix. For example, when you use Prisma, you supposed to import it from another file where it's already initialized. What I would like to do is something like this:

```js
const appFactory = (deps) => {
  const prismaClient = prismaFactory(deps);
  const expressServer = expressServerFactory({ ...deps, prismaClient });
  
  return {
    run: () => expressServer.listen()
  };
}

const app = appFactory({ /* ... pass all deps for the entire app */ });
app.run();
```

The `expressServerFactory` function would also define all the routes and pass `prismaClient` to each route handler factory function as a dependency. This way, I have zero global variables and code that is super easy to test, mock and replace modules without affecting the rest of the code.

However, the routes convention used in Remix kinda makes it impossible. I've seen that you can define your routes in the config file but it's outside of the `src` directory so I can't use this pattern.

Using global variables maybe is easy and tempting to use but it's programming antipattern.

Do you guys know how to apply this pattern in Remix?

**Comment by sergiodxa** [comment]
If you are using Express you can pass a context object to the loaders and actions on every request from the Express server using the getLoadContext, see https://remix.run/docs/en/v1/other-api/adapter#createrequesthandler, this context is accesible in any loader or action.

**Reply by lukejagodzinski** [reply]
Yep, I've seen it but there is no full example of how to use it. I was thinking that it might work as you described but haven't yet tested. Actually, in my tests I'm not using express. I'm using Remix server. Is it also possible with Remix server?

**Reply by sergiodxa** [reply]
The Remix App Server uses Express but it doesn’t let you control the server code. But you can change it to be Express and control it.

**Reply by lukejagodzinski** [reply]
Ok cool, I will try that. Thanks!

**Reply by lukejagodzinski** [reply]
It's actually really bad solution. The `server/index.js` file is written in JS and I want to share code written in TS. I think the best thing would be to have more control over the context object in the Remix server.

**Reply by ccssmnn** [reply]
You can write it in TS and add a separate build step to generate the final JS file to start the server.

**Reply by lukejagodzinski** [reply]
I could, but should I? If the Remix framework supports TS, then it would be nice if it also supported it in config/setup files.

**Reply by ccssmnn** [reply]
Remix is TS. What you build is a [Remix HTTP Handler](https://remix.run/docs/en/v1/pages/technical-explanation#http-handler-and-adapters) that you can use. Remix ends here. You then use adapters (or implement your own) to plug in the Remix HTTP Handler into a server framework of your choice. That could be a Vercel Function, a Cloudflare Worker, an Express server. But that is not Remix anymore. 

In the case of Express, you use the Remix HTTP Handler via the Express **Adapter**. When you want to build your express server with TS, you are free to do so. Build some of your routes with Express, handle other ones or everything else with the Remix Handler. It's completely your choice. You may need to change the Remix server build, to match the compiled output of your express implementation.
